packages/messaging/test/sw-controller.test.ts

```diff
@@ -638,162 +638,164 @@ describe('Firebase Messaging > *SwController', () => {
       expect(event.notification.close.callCount).to.equal(1);
     });
 
-    it('should open click_action URL for FCM notification (third-party so no window client access)', async () => {
-      const clickAction = '/test-click-action';
-      const waitUntilSpy = sandbox.spy();
-      const event: any = {
-        notification: {
-          data: {
-            FCM_MSG: {
+    ['link', 'click_action'].forEach(key => {
+      describe(key, () => {
+        let link: string;
+        let waitUntilSpy: sinon.SinonSpy;
+        let event: any;
+
+        beforeEach(() => {
+          link = '/test-link';
+          waitUntilSpy = sandbox.spy();
+          if (key === 'link') {
+            // New API
+            event = {
               notification: {
-                click_action: clickAction
-              }
-            }
-          },
-          close: sandbox.spy()
-        },
-        waitUntil: waitUntilSpy,
-        stopImmediatePropagation: sandbox.spy()
-      };
+                data: {
+                  FCM_MSG: {
```

Comment:
nit: since these two paths differ only below FCM_MSG, pull it out? Alternatively, consider a helper method that just builds the event given the key?

const msg = key === 'link'
    ? {notification: {}, fcmOptions: {link}}
    : {notification: {click_action: link}};

event = {
  notification: {
    data: {FCM_MSG: msg},
    close: sandbox.spy()
  },
  waitUntil: waitUntilSpy,
  stopImmediatePropagation: sandbox.spy()
};

Reply:
I did consider the helper method but I'd like to keep logic in tests as simple as possible. Still, I guess I can pull FCM_MSG out...